File: example/scalalib/web/6-webapp-scalajs-shared/client/src/ClientApp.scala
package client
import org.scalajs.dom
import shared.{Todo, Shared}
object ClientApp {
  var state = "all"
  var todoApp = dom.document.getElementsByClassName("todoapp")(0)

  def postFetchUpdate(url: String) = {
    dom.fetch(
      url,
      new dom.RequestInit {
        method = dom.HttpMethod.POST
      }
    ).`then`[String](response => response.text())
      .`then`[Unit] { text =>
        todoApp.innerHTML = Shared
          .renderBody(upickle.default.read[Seq[Todo]](text), state)
          .render

        initListeners()
      }
  }

  def bindEvent(cls: String, url: String, endState: Option[String]) = {
    dom.document.getElementsByClassName(cls)(0).addEventListener(
      "mousedown",
      (evt: dom.Event) => {
        postFetchUpdate(url)
        endState.foreach(state = _)
      }
    )
  }

  def bindIndexedEvent(cls: String, func: String => String) = {
    for (elem <- dom.document.getElementsByClassName(cls)) {
      elem.addEventListener(
        "mousedown",
        (evt: dom.Event) => postFetchUpdate(func(elem.getAttribute("data-todo-index")))
      )
    }
  }

  def initListeners(): Unit = {
    bindIndexedEvent("destroy", index => s"/delete/$state/$index")
    bindIndexedEvent("toggle", index => s"/toggle/$state/$index")
    bindEvent("toggle-all", s"/toggle-all/$state", None)
    bindEvent("todo-all", s"/list/all", Some("all"))
    bindEvent("todo-active", s"/list/all", Some("active"))
    bindEvent("todo-completed", s"/list/completed", Some("completed"))
    bindEvent("clear-completed", s"/clear-completed/$state", None)

    val newTodoInput =
      dom.document.getElementsByClassName("new-todo")(0).asInstanceOf[dom.HTMLInputElement]
    newTodoInput.addEventListener(
      "keydown",
      (evt: dom.KeyboardEvent) => {
        if (evt.keyCode == 13) {
          dom.fetch(
            s"/add/$state",
            new dom.RequestInit {
              method = dom.HttpMethod.POST
              body = newTodoInput.value
            }
          ).`then`[String](response => response.text())
            .`then`[Unit] { text =>
              newTodoInput.value = ""

              todoApp.innerHTML = Shared
                .renderBody(upickle.default.read[Seq[Todo]](text), state)
                .render

              initListeners()
            }
        }
      }
    )
  }

  def main(args: Array[String]): Unit = initListeners()
}

File: example/kotlinlib/web/4-webapp-kotlinjs/client/src/ClientApp.kt
package client

import kotlinx.browser.document
import kotlinx.browser.window
import org.w3c.dom.Element
import org.w3c.dom.HTMLInputElement
import org.w3c.dom.asList
import org.w3c.dom.events.KeyboardEvent
import org.w3c.dom.get
import org.w3c.fetch.RequestInit

object ClientApp {
    private var state = "all"

    private val todoApp: Element
        get() = checkNotNull(document.getElementsByClassName("todoapp")[0])

    private fun postFetchUpdate(url: String) {
        window
            .fetch(url, RequestInit(method = "POST"))
            .then { it.text() }
            .then { text ->
                todoApp.innerHTML = text
                initListeners()
            }
    }

    private fun bindEvent(
        cls: String,
        url: String,
        endState: String? = null,
    ) {
        document
            .getElementsByClassName(cls)[0]
            ?.addEventListener(
                "click",
                {
                    postFetchUpdate(url)
                    if (endState != null) state = endState
                },
            )
    }

    private fun bindIndexedEvent(
        cls: String,
        block: (String) -> String,
    ) {
        for (elem in document.getElementsByClassName(cls).asList()) {
            elem.addEventListener(
                "click",
                { postFetchUpdate(block(elem.getAttribute("data-todo-index")!!)) },
            )
        }
    }

    fun initListeners() {
        bindIndexedEvent("destroy") {
            "/delete/$state/$it"
        }
        bindIndexedEvent("toggle") {
            "/toggle/$state/$it"
        }
        bindEvent("toggle-all", "/toggle-all/$state")
        bindEvent("todo-all", "/list/all", "all")
        bindEvent("todo-active", "/list/active", "active")
        bindEvent("todo-completed", "/list/completed", "completed")
        bindEvent("clear-completed", "/clear-completed/$state")

        val newTodoInput = document.getElementsByClassName("new-todo")[0] as HTMLInputElement
        newTodoInput.addEventListener(
            "keydown",
            {
                check(it is KeyboardEvent)
                if (it.keyCode == 13) {
                    window
                        .fetch("/add/$state", RequestInit(method = "POST", body = newTodoInput.value))
                        .then { it.text() }
                        .then { text ->
                            newTodoInput.value = ""
                            todoApp.innerHTML = text
                            initListeners()
                        }
                }
            },
        )
    }
}

fun main(args: Array<String>) = ClientApp.initListeners()
